include/debug/coredump.h

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2020 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DEBUG_COREDUMP_H_
+#define ZEPHYR_INCLUDE_DEBUG_COREDUMP_H_
+
+#ifdef CONFIG_DEBUG_COREDUMP
+
+#include <toolchain.h>
+#include <arch/cpu.h>
+#include <sys/byteorder.h>
+
+#define Z_COREDUMP_HDR_VER		1
+
+#define	Z_COREDUMP_ARCH_HDR_ID		'A'
+
+#define	Z_COREDUMP_MEM_HDR_ID		'M'
+#define Z_COREDUMP_MEM_HDR_VER		1
+
+/* Target code */
+enum z_coredump_tgt_code {
+	COREDUMP_TGT_UNKNOWN = 0,
+};
+
+/* Coredump header */
+struct z_coredump_hdr_t {
+	/* 'Z', 'E' */
+	char		id[2];
+
+	/* Header version */
+	uint16_t	hdr_version;
+
+	/* Target code */
+	uint16_t	tgt_code;
+
+	/* Pointer size in Log2 */
+	uint8_t		ptr_size_bits;
+
+	uint8_t		flag;
+
+	/* Coredump Reason given */
+	unsigned int	reason;
+} __packed;
+
+/* Architecture-specific block header */
+struct z_coredump_arch_hdr_t {
+	/* Z_COREDUMP_ARCH_HDR_ID */
+	char		id;
+
+	/* Header version */
+	uint16_t	hdr_version;
+
+	/* Number of bytes in this block (excluding header) */
+	uint16_t	num_bytes;
+} __packed;
+
+/* Memory block header */
+struct z_coredump_mem_hdr_t {
+	/* Z_COREDUMP_MEM_HDR_ID */
+	char		id;
+
+	/* Header version */
+	uint16_t	hdr_version;
+
+	/* Address of start of memory region */
+	uintptr_t	start;
+
+	/* Address of end of memory region */
+	uintptr_t	end;
+} __packed;
+
+void z_coredump(unsigned int reason, const z_arch_esf_t *esf,
+		struct k_thread *thread);
+void z_coredump_memory_dump(uintptr_t start_addr, uintptr_t end_addr);
+int z_coredump_buffer_output(uint8_t *buf, size_t buflen);
+
+#else
+
+void z_coredump(unsigned int reason, const z_arch_esf_t *esf,
+		struct k_thread *thread)
+{
+}
+
+void z_coredump_memory_dump(uintptr_t start_addr, uintptr_t end_addr)
+{
+}
+
+int z_coredump_buffer_output(uint8_t *buf, size_t buflen)
+{
+	return 0;
+}
+
+#endif /* CONFIG_DEBUG_COREDUMP */
+
+/**
+ * @defgroup coredump_apis Coredump APIs
+ * @brief Coredump APIs
+ * @{
+ */
+
+/**
+ * @fn void z_coredump(unsigned int reason, const z_arch_esf_t *esf, struct k_thread *thread);
+ * @brief Perform coredump during z_fatal_error().
+ *
+ * This should be called with z_fatal_error() to generate coredump.
```

Comment:
similar to a comment @andrewboie did in my gdbstub pr, shouldn't this API be in a private header ?

Reply:
yeah if this is really intended for `z_fatal_error()` it should be private.

but if you actually meant for it to be called from an implementation of `k_sys_fatal_error_handler()` (which is customizable) then it can remain and the comment should be updated

Reply:
The idea is that this can also be called on demand outside of fatal errors if one is determined to do so. GDB allows you to do core dump while debugging, so this is (somehow) mirroring that ability. Therefore it is under `include/` but has a prefix `z_`.